syntax = "proto3";

service Chatting {
  rpc ConnectServer(ConnectServerReqeust) returns (ConnectServerResponse);
  rpc ListenChat(ListenRequest) returns (stream ChatResponse);
  rpc SendChat(ChatRequest) returns (Empty);
}

message Empty {}

message ListenRequest {
  //  string room_name = 1;
  //  UserInfo sender = 3;
  reserved 1, 3;
  UserInfo me = 2;
}

message ConnectServerReqeust {
  string user_name = 1;
}

message ConnectServerResponse {
  bool success = 1;
  oneof user_or_error {
    UserInfo user = 2;
    string error_message = 3;
  }
}

message UserInfo {
  string user_name = 1;
  optional string random_color = 3;
  //  string uid = 2;
  reserved 2;
}

message ChatRequest {
  UserInfo sender = 1;
  oneof msg_or_file {
    string message = 2;
    //    File file = 6;
  }
  //  google.protobuf.Timestamp timestamp = 3;
  //  string room_name = 4;
  reserved 3, 4, 5, 6;
  //  UserInfo target = 5;
}

message ChatResponse {
  bool success = 1;
  oneof req_or_error {
    ChatRequest chat_req = 2;
    string error_message = 3; // Check file(6)'s size is not over 2^32.
  }
}
